ada_feeding_perception/ada_feeding_perception/face_detection.py

```diff
@@ -0,0 +1,388 @@
+                        description="The name of the face detection model checkpoint to use",
+                        read_only=True,
+                    ),
+                ),
+                (
+                    "landmark_model_base_url",
+                    None,
+                    ParameterDescriptor(
+                        name="landmark_model_base_url",
+                        type=ParameterType.PARAMETER_STRING,
+                        description=(
+                            "The URL to download the model checkpoint from "
+                            "if it is not already downloaded"
+                        ),
+                        read_only=True,
+                    ),
+                ),
+                (
+                    "model_dir",
+                    None,
+                    ParameterDescriptor(
+                        name="model_dir",
+                        type=ParameterType.PARAMETER_STRING,
+                        description=(
+                            "The location of the directory where the model "
+                            "checkpoint is / should be stored"
+                        ),
+                        read_only=True,
+                    ),
+                ),
+            ],
+        )
+
+    def toggle_face_detection_callback(
+        self, request: SetBool.Request, response: SetBool.Response
+    ) -> SetBool.Response:
+        """
+        Callback function for the toggle_face_detection service. Safely toggles
+        the face detection on or off depending on the request.
+        """
+        self.get_logger().info(f"Incoming service request. data: {request.data}")
+        with self.is_on_lock:
+            self.is_on = request.data
+        response.success = request.data
+        return response
+
+    def camera_info_callback(self, msg: CameraInfo):
+        """
+        Callback function for the toggle_face_detection service. Safely toggles
+        the face detection on or off depending on the request.
+        """
+        self.camera_matrix = msg.k
+
+    def depth_callback(self, msg: CompressedImage):
+        """
+        Callback function for depth images. If face_detection is on, this
+        function publishes the 3d location of the mouth on the /face_detection
+        topic in the camera_depth_optical_frame
+        """
+
+        # Collect last ten depth images (depth_buffer is a collections.deque of max length 10)
+        self.depth_buffer.append(msg)
+        with self.is_on_lock:
+            is_on = self.is_on
+
+        # Check if face detection is on
+        if is_on:
+            face_detection_msg = FaceDetection()
+            if self.is_face_detected:
+                # Retrieve the 2d location of the mouth center
+                u = int(self.img_mouth_center[0])
+                v = int(self.img_mouth_center[1])
+
+                # Find depth image closest in time to saved color image
+                difference_array = np.array([])
```

Comment:
Nit: since you know the length of `difference_array` before-hand (the same length as `depth_img`) you should instead create an array of that size (e.g., `difference_array = np.zeros((len(self.depth_buffer),), stype=np.float32)`) and then iterate over the indices (`for i in range(len(self.depth_buffer)`) and set that index of the difference array instead of appending (`difference_array[i] = abs(depth_time - img_time)`). This prevents multiple changes of the size of the underlying array.